src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,243 @@
+        }
+
+        auto& target = *table.get_link_target(ColKey(link.col_key));
+        auto lvr = obj.get_linklist(ColKey(link.col_key));
+        return std::any_of(lvr.begin(), lvr.end(), [&, this](auto key) {
+            return this->check_row(target, key.value, filtered_columns, depth + 1);
+        });
+    };
+
+    // Check the `links` of all `m_related_tables` and return true if any of them has a `linked_object_changed`.
+    return std::any_of(begin(it->links), end(it->links), linked_object_changed);
+}
+
+bool DeepChangeChecker::check_row(Table const& table, ObjKeyType key, std::vector<ColKey> filtered_columns,
+                                  size_t depth)
+{
+    // Arbitrary upper limit on the maximum depth to search
+    if (depth >= m_current_path.size()) {
+        // Don't mark any of the intermediate rows checked along the path as
+        // not modified, as a search starting from them might hit a modification
+        for (size_t i = 0; i < m_current_path.size(); ++i)
+            m_current_path[i].depth_exceeded = true;
+        return false;
+    }
+
+    TableKey table_key = table.get_key();
+
+    // If the pair (table_key.value, key) can be found in `m_info.tables` we can
+    // end the search and return here.
+    if (depth > 0) {
+        auto it = m_info.tables.find(table_key.value);
+        if (it != m_info.tables.end() && it->second.modifications_contains(key, filtered_columns))
+            return true;
+    }
+
+    // Look up the unmodified objects for the `table_key.value` and check if the
+    // `key` can be found within them. If so, we can return without checking the
+    // outgoing links.
+    auto& not_modified = m_not_modified[table_key.value];
+    auto it = not_modified.find(key);
+    if (it != not_modified.end())
+        return false;
+
+    // If both of the above short cuts don't lead to a result we need to check the
+    // outgoing links.
+    bool ret = check_outgoing_links(table_key, table, key, filtered_columns, depth);
+    if (!ret && (depth == 0 || !m_current_path[depth - 1].depth_exceeded))
+        not_modified.insert(key);
+    return ret;
+}
+
+bool DeepChangeChecker::operator()(ObjKeyType key)
+{
+    std::vector<ColKey> filtered_columns_in_root_table = {};
+    std::vector<ColKey> filtered_columns = {};
+
+    // If all callbacks do have a filter, every `KeyPathArray` will have entries.
+    // In this case we need to check the `ColKey`s and pass the filtered columns
+    // to the checker.
+    // If at least one `Callback` does not have a filter we notify on any change.
+    // This is signaled by leaving the `filtered_columns_in_root_table` and
+    // `filtered_columns` empty.
+    auto all_callbacks_filtered = all_of(begin(m_key_path_arrays), end(m_key_path_arrays), [](auto key_path_array) {
+        return key_path_array.size() > 0;
+    });
+    if (all_callbacks_filtered) {
```

Comment:
All of this should be done when constructing the DeepChangeChecker and not for each row checked.

Reply:
Changed with 8fd4b1ec5. However, I think I do have the same problem with non-uniqueness of `ColKey` here and need to rework this further.

Reply:
This is now addressed with 0cf4086ae.

I somehow think we could just move the two checks for `if (!any_related_table_was_modified(info)) {` and `if (m_related_tables.size() == 1 && !all_callbacks_filtered()) {` into the `DeepChangeChecker` as well and then always return a `DeepChangeChecker` which takes care of any change checking and with that substitute `std::function<bool(ObjectChangeSet::ObjectKeyType)>` on `get_modification_checker` with `DeepChangeChecker`.